File: engine/network/edge.py
"""Road network edges (road segments) with dynamic traffic state."""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum


class RoadType(StrEnum):
    STREET = "street"
    AVENUE = "avenue"
    HIGHWAY = "highway"


class RoadStatus(StrEnum):
    OPEN = "open"
    CLOSED = "closed"
    ACCIDENT = "accident"
    CONSTRUCTION = "construction"


@dataclass
class Road:
    """A directed road segment between two nodes.

    Units: ``distance`` is in metres, ``speed_limit`` in metres/second
    (scenario files express limits in km/h; the builder converts them).

    Dynamic state (``current_vehicle_count``, ``congestion``,
    ``current_travel_time``) is recomputed each simulation tick.
    """

    id: str
    source: str
    destination: str
    distance: float
    speed_limit: float
    capacity: int
    lanes: int = 1
    road_type: RoadType = RoadType.STREET
    status: RoadStatus = RoadStatus.OPEN
    priority: float = 0.0
    current_vehicle_count: int = 0
    congestion: float = 0.0
    base_travel_time: float = 0.0
    current_travel_time: float = 0.0

    def __post_init__(self) -> None:
        if self.distance <= 0:
            raise ValueError(f"road {self.id}: distance must be positive")
        if self.speed_limit <= 0:
            raise ValueError(f"road {self.id}: speed_limit must be positive")
        if self.capacity <= 0:
            raise ValueError(f"road {self.id}: capacity must be positive")
        self.base_travel_time = self.distance / self.speed_limit
        self.current_travel_time = self.base_travel_time

    @property
    def is_traversable(self) -> bool:
        return self.status is RoadStatus.OPEN

    def to_dict(self) -> dict[str, object]:
        return {
            "id": self.id,
            "source": self.source,
            "destination": self.destination,
            "distance": self.distance,
            "speed_limit": self.speed_limit,
            "capacity": self.capacity,
            "lanes": self.lanes,
            "road_type": self.road_type.value,
            "status": self.status.value,
            "current_vehicle_count": self.current_vehicle_count,
            "congestion": self.congestion,
            "base_travel_time": self.base_travel_time,
            "current_travel_time": self.current_travel_time,
        }

File: browser-engine/engine/network/edge.ts
import { Road, RoadStatus, RoadType } from '../shared/types';

export function createRoad(params: {
  id: string;
  source: string;
  destination: string;
  distance: number;
  speedLimit: number;
  capacity: number;
  lanes?: number;
  roadType?: RoadType;
  status?: RoadStatus;
  priority?: number;
}): Road {
  const {
    id,
    source,
    destination,
    distance,
    speedLimit,
    capacity,
    lanes = 1,
    roadType = RoadType.STREET,
    status = RoadStatus.OPEN,
    priority = 0,
  } = params;

  if (distance <= 0) throw new Error(`road ${id}: distance must be positive`);
  if (speedLimit <= 0) throw new Error(`road ${id}: speedLimit must be positive`);
  if (capacity <= 0) throw new Error(`road ${id}: capacity must be positive`);

  const baseTravelTime = distance / speedLimit;

  return {
    id,
    source,
    destination,
    distance,
    speedLimit,
    capacity,
    lanes,
    roadType,
    status,
    priority,
    currentVehicleCount: 0,
    congestion: 0,
    baseTravelTime,
    currentTravelTime: baseTravelTime,
  };
}

export function roadToDict(road: Road): Record<string, unknown> {
  return {
    id: road.id,
    source: road.source,
    destination: road.destination,
    distance: road.distance,
    speedLimit: road.speedLimit,
    capacity: road.capacity,
    lanes: road.lanes,
    roadType: road.roadType,
    status: road.status,
    currentVehicleCount: road.currentVehicleCount,
    congestion: road.congestion,
    baseTravelTime: road.baseTravelTime,
    currentTravelTime: road.currentTravelTime,
  };
}

export function isTraversable(road: Road): boolean {
  return road.status === RoadStatus.OPEN;
}

export function congestionFactor(congestion: number): number {
  if (congestion <= 0.3) return 1.0;
  if (congestion <= 0.7) return 1.0 + 2.0 * (congestion - 0.3);
  return 1.8 + 10.0 * (congestion - 0.7);
}

export function updateRoadDynamicState(road: Road, vehicleCount: number): void {
  road.currentVehicleCount = vehicleCount;
  road.congestion = road.capacity > 0 ? vehicleCount / road.capacity : 1.0;
  road.currentTravelTime = road.baseTravelTime * congestionFactor(road.congestion);
}